diesel/src/pg/connection/cursor.rs

```diff
@@ -24,13 +28,13 @@ impl<'a> ExactSizeIterator for Cursor<'a> {
 }
 
 impl<'a> Iterator for Cursor<'a> {
-    type Item = PgRow<'a>;
+    type Item = crate::QueryResult<PgRow<'a>>;
```

Comment:
It looks like we don't really need `Iterator` that returns `Row`s: all we ever do with it is call `.map(FromSqlRow::build_from_row)`. If we instead make an iterator that both creates the `Row<'a>` and calls `FromSqlRow` on it as part of a single iterator we don't need to add `Rc`s everywhere.

That being said in row-by-row mode there's no need to keep a reference to a global result object, the `PGresult` returned by each call to `PQgetResult` is enough so we could just store it in `Row`. ([doc](https://www.postgresql.org/docs/current/libpq-single-row-mode.html))

Reply:
> It looks like we don't really need Iterator that returns Rows: all we ever do with it is call .map(FromSqlRow::build_from_row). If we instead make an iterator that both creates the Row<'a> and calls FromSqlRow on it as part of a single iterator we don't need to add Rcs everywhere. 

For diesel itself it's true that we never do anything else than calling `FromSqlRow::build_from_row` onto the returned row. The idea behind this change is that it's now possible to just skip diesels deserialization layer, in such a way that it is possible to interact with the raw result set. This is useful for cases where you don't know the schema at compile time and want to build for example select clauses at runtime. In the end this just tries to separate different diesel layers more strictly, so that in the future someone can decide to use diesels `Connection` but to skip the query builder entirely. 

That all written, I'm not sure if that's even worth it and if it will work with all backends.